include/deal.II/lac/trilinos_tpetra_solver_direct.templates.h

```diff
@@ -0,0 +1,225 @@
+
+#ifndef dealii_trilinos_tpetra_solver_templates_h
+#define dealii_trilinos_tpetra_solver_templates_h
+
+#include <deal.II/base/config.h>
+
+#include "deal.II/lac/solver_control.h"
+
+#include <string>
+
+
+#ifdef DEAL_II_TRILINOS_WITH_TPETRA
+#  ifdef DEAL_II_TRILINOS_WITH_AMESOS2
+
+#    include "deal.II/base/template_constraints.h"
+#    include <deal.II/base/conditional_ostream.h>
+
+#    include <deal.II/lac/trilinos_tpetra_solver_direct.h>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace LinearAlgebra
+{
+  namespace TpetraWrappers
+  {
+    /* ---------------------- SolverDirectBase ------------------------ */
+
+    template <typename Number, typename MemorySpace>
+    SolverDirectBase<Number, MemorySpace>::SolverDirectBase(
+      SolverControl     &cn,
+      const std::string &solver_type,
+      const bool         output_solver_details)
+      : solver_control(cn)
+      , solver_type(solver_type)
+      , output_solver_details(output_solver_details)
+    {
+      AssertThrow(Amesos2::query(solver_type),
+                  ExcTrilinosAmesos2SolverUnsupported(solver_type));
+    }
+
+    template <typename Number, typename MemorySpace>
+    SolverControl &
+    SolverDirectBase<Number, MemorySpace>::control() const
+    {
+      return solver_control;
+    }
+
+    template <typename Number, typename MemorySpace>
+    void
+    SolverDirectBase<Number, MemorySpace>::initialize(
+      const SparseMatrix<Number, MemorySpace> &A)
+    {
+      // First set whether we want to print the solver information to screen or
+      // not.
+      ConditionalOStream verbose_cout(std::cout, output_solver_details);
+
+      // Next allocate the Amesos2 solver with the concrete solver, if possible.
+      solver =
+        Amesos2::create<typename SparseMatrix<Number, MemorySpace>::MatrixType,
+                        MultiVector>(solver_type, A.trilinos_rcp());
+
+      solver->setParameters(Teuchos::rcpFromRef(parameter_list));
+      // Now do the actual factorization, which is a two step procedure.
+      // The symbolic factorization determines the structure of the inverse,
+      // while the numeric factorization does to actual computation of L and U
+      verbose_cout << "Starting symbolic factorization" << std::endl;
+      solver->symbolicFactorization();
+
+      verbose_cout << "Starting numeric factorization" << std::endl;
+      solver->numericFactorization();
+    }
+
+    template <typename Number, typename MemorySpace>
+    void
+    SolverDirectBase<Number, MemorySpace>::solve(
+      Vector<Number, MemorySpace>       &x,
+      const Vector<Number, MemorySpace> &b)
+    {
+      // Assign the empty solution vector
+      solver->setX(x.trilinos_rcp());
+
+      // Assign the RHS vector
+      solver->setB(b.trilinos_rcp());
+      // First set whether we want to print the solver information to screen
+      // or not.
+      ConditionalOStream verbose_cout(std::cout, output_solver_details);
```

Comment:
Same here and elsewhere.

Reply:
I am not sure if this relates to the three lines rule or if you meant anything else?

Reply:
This related to a comment that must have gone lost. I don't think the `verbose_cout` thing is very useful. Let's use the opportunity to get rid of it here and in all of the other places where we use it in these direct solvers, along with the flag that controls whether to output.